### PR TITLE
Introduce ResourceStoreFactory

### PR DESCRIPTION
**Pull request by dvntucker**
Changes to how ResourceStore is loaded and removing ResourceStore interface from classes where it didn't make much sense. Mainly to facilitate overriding the store implementation. 

- ResourceStoreProxy is now ResourceStoreFactory. Factory first searches for a "resourceStoreImpl" bean before falling back to the data directory implementation.

- Removed ResourceStore interface from GeoServerSecurityManager, I don't think it makes sense for this class to be a store and it wasn't really being used that way anyway.

- Removed ResourceStore interface from GeoServerDataDirectory where it wasn't really being used.

### CODE REVIEW

File: src/community/jdbcconfig/src/test/java/org/geoserver/jdbcconfig/internal/JDBCConfigPropertiesTest.java [hunk 1]
```diff
@@ -17,6 +12,15 @@
...
 import org.junit.Test;
 
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.Properties;
+
+import static org.geoserver.jdbcconfig.JDBCConfigTestSupport.createTempDir;
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.junit.Assert.*;
```
Comment (jodygarnett) [inline src/community/jdbcconfig/src/test/java/org/geoserver/jdbcconfig/internal/JDBCConfigPropertiesTest.java]: Is this just a change of imports? Wonder if your IDE is sorting things in a different order than eclipse?

File: src/community/jdbcstore/src/main/resources/applicationContext.xml [hunk 2]
```diff
@@ -25,7 +19,7 @@
   </bean>
   
   <!-- resource store -->
-  <bean id="jdbcResourceStore" class="org.geoserver.jdbcstore.JDBCResourceStoreFactoryBean">  
+  <bean id="resourceStoreImpl" class="org.geoserver.jdbcstore.JDBCResourceStoreFactoryBean">
```
Comment (jodygarnett) [inline src/community/jdbcstore/src/main/resources/applicationContext.xml]: See what you mean about "resourceStoreImpl", can you call it resourceStoreFactory?

It feels like we are doing something wrong when we have to declare a factory, but yeah spring factory bean is apparently the way to go. I cannot see a consistent readable naming convention for factory beans.

File: src/main/src/main/java/org/geoserver/config/GeoServerDataDirectory.java [hunk 3]
```diff
@@ -81,7 +54,7 @@
  * @author Justin Deoliveira, OpenGeo
  */
 @SuppressWarnings("unused")
-public class GeoServerDataDirectory implements ResourceStore {
+public class GeoServerDataDirectory {
```
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/config/GeoServerDataDirectory.java]: I am really excited you were able to do this, excellent work @NielsCharlier for making this possible.

It is an API change, but I guess that is what we are up to here, trying to factor out things correctly so code can consistently get at a ResourceStore.

File: src/main/src/main/java/org/geoserver/security/GeoServerSecurityManager.java [hunk 4]
```diff
@@ -164,13 +68,26 @@
...
+import java.security.InvalidKeyException;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.*;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static org.geoserver.data.util.IOUtils.xStreamPersist;
```
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/security/GeoServerSecurityManager.java]: Really feels like your IDE is sorting differently, can we check (if not the next eclipse developer will undo your work here)

File: src/platform/src/main/java/org/geoserver/platform/resource/ResourceStoreFactory.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.FactoryBean;
+import org.springframework.beans.factory.NoSuchBeanDefinitionException;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class ResourceStoreFactory implements FactoryBean<ResourceStore>, ApplicationContextAware {
```
Comment (jodygarnett) [inline src/platform/src/main/java/org/geoserver/platform/resource/ResourceStoreFactory.java]: class javadoc, yeah I know right.

File: src/main/src/main/java/org/geoserver/security/GeoServerSecurityManager.java (hunk 4 above)
Comment (dvntucker) [inline src/main/src/main/java/org/geoserver/security/GeoServerSecurityManager.java]: I have updated IntelliJ to organize imports like Eclipse.

**Comment by jodygarnett** [comment]
Thanks for the QA & Testing @dvntucker - if we can clean up the imports I would like to get this change in (in part because it is an API change for work introduced this cycle).

File: src/main/src/main/java/applicationContext.xml [hunk 6]
```diff
@@ -22,10 +22,11 @@
 	<!-- resources -->
     <bean id="dataDirectoryResourceStore" class="org.geoserver.platform.resource.DataDirectoryResourceStore">
         <property name="lockProvider" ref="lockProvider"/>
-    </bean>  
-    <bean id="resourceStore" class="org.geoserver.platform.resource.ResourceStoreProxy">
-        <property name="delegate" ref="dataDirectoryResourceStore"/>
     </bean>
+
+    <bean id="resourceStore" class="org.geoserver.platform.resource.ResourceStoreFactory"
+          depends-on="dataDirectoryResourceStore"/>
```
Comment (jodygarnett) [inline src/main/src/main/java/applicationContext.xml]: what is this for

**Comment by jodygarnett** [comment]
When trying to merge/rebase I get a test failure in app-schema:

````
-------------------------------------------------------------------------------
Test set: org.geoserver.test.LocalResolveTest
-------------------------------------------------------------------------------
Tests run: 6, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.805 sec <<< FAILURE!
testResolveTimeOut(org.geoserver.test.LocalResolveTest)  Time elapsed: 43 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[urn:x-test:GeologicUnit:gu.25699]> but was:<[]>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.geoserver.test.AbstractAppSchemaTestSupport.assertXpathEvaluatesTo(AbstractAppSchemaTestSupport.java:344)
	at org.geoserver.test.LocalResolveTest.testResolveTimeOut(LocalResolveTest.java:106)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
````

Any thoughts @bencaradocdavies ?

**Comment by jodygarnett** [comment]
tested master again (works) and this branch again ... getting a different failure:

````
Tests run: 6, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.867 sec <<< FAILURE!
testImportBuildingsCancellation(org.geoserver.wps.gs.ImportProcessTest)  Time elapsed: 11 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[Test messag]e> but was:<[Failed to import data into the target stor]e>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.geoserver.wps.gs.ImportProcessTest.testImportBuildingsCancellation(ImportProcessTest.java:149)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
````

Could we be looking at multi-threading issue? I am using ``mvn install -T 2C -Prelease`` to test the branch since tycho is MIA.

**Comment by jodygarnett** [comment]
Dropping back to ``mvn clean install -Prelease`` fixed things ... not sure how to proceed going to focus on other pull requests.

**Comment by jodygarnett** [comment]
merged to master

**Comment by bencaradocdavies** [comment]
The failure in testResolveTimeOut(org.geoserver.test.LocalResolveTest) is intermittent; I have not seen this for a while but IIRC this test involves a thread and a timeout; we used to see a failure from time to time.

**Pull request closed without merging**